### ISSUE TITLE
Support for PHP 8.1

### ISSUE THREAD

**Issue by sebastic** [Icinga/icingaweb2#4609]
## Describe the bug
The PHP 8 support in Icingaweb2 is not sufficient to work with PHP 8.1.

Accessing `/icingaweb2` results in:
```
Deprecated: Return type of Icinga\Application\Libraries::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Libraries.php on line 20

Deprecated: Return type of Icinga\Application\Config::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 125

Deprecated: Return type of Icinga\Application\Config::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 145

Deprecated: Return type of Icinga\Application\Config::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 175

Deprecated: Return type of Icinga\Application\Config::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 165

Deprecated: Return type of Icinga\Application\Config::valid() should either be compatible with Iterator::valid(): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 155

Deprecated: Return type of Icinga\Application\Config::rewind() should either be compatible with Iterator::rewind(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 135

Deprecated: Return type of Icinga\Web\Session\SessionNamespace::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Web/Session/SessionNamespace.php on line 35

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 91

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::offsetExists($priorityOrHelperName) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 102

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::offsetGet($priorityOrHelperName) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 117

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::offsetSet($priority, $helper) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 137

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::offsetUnset($priorityOrHelperName) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 173

Deprecated: Return type of Zend_Controller_Action_HelperBroker_PriorityStack::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php on line 198

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 317

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 261

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 273

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::offsetGet($offset) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 284

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::offsetSet($offset, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 296

Deprecated: Return type of Zend_View_Helper_Placeholder_Container_Standalone::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/View/Helper/Placeholder/Container/Standalone.php on line 307

Deprecated: Return type of Zend_Paginator::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Paginator.php on line 513

Deprecated: Return type of Zend_Paginator::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/icingaweb2/library/vendor/Zend/Paginator.php on line 815

Deprecated: Return type of Icinga\Authentication\AuthChain::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Authentication/AuthChain.php on line 200

Fatal error: Uncaught ErrorException: Uncaught Zend_Controller_Response_Exception: Cannot send headers; headers already sent in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php, line 30 in /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php:321 Stack trace: #0 /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php(148): Zend_Controller_Response_Abstract->canSendHeaders() #1 /usr/share/php/Icinga/Web/Response.php(356): Zend_Controller_Response_Abstract->setRedirect() #2 /usr/share/php/Icinga/Web/Response.php(407): Icinga\Web\Response->prepare() #3 /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php(762): Icinga\Web\Response->sendHeaders() #4 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(975): Zend_Controller_Response_Abstract->sendResponse() #5 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch() #6 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch() #7 /usr/share/icingaweb2/public/index.php(4): require_once('...') #8 {main} thrown in /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php:321 Stack trace: #0 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}() #1 {main} Next Zend_Controller_Exception: Uncaught Zend_Controller_Response_Exception: Cannot send headers; headers already sent in /usr/share/icingaweb2/library/vendor/Zend/Controller/Action/HelperBroker/PriorityStack.php, line 30 in /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php:321 Stack trace: #0 /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php(148): Zend_Controller_Response_Abstract->canSendHeaders() #1 /usr/share/php/Icinga/Web/Response.php(356): Zend_Controller_Response_Abstract->setRedirect() #2 /usr/share/php/Icinga/Web/Response.php(407): Icinga\Web\Response->prepare() #3 /usr/share/icingaweb2/library/vendor/Zend/Controller/Response/Abstract.php(762): Icinga\Web\Response->sendHeaders() #4 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(975): Zend_Controller_Response_Abstract->sendResponse() #5 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch() #6 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch() #7 /usr/share/icingaweb2/public/index.php(4): require_once('...') #8 {main} thrown#0 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}() #1 {main} in /usr/share/icingaweb2/library/vendor/Zend/Controller/Plugin/Broker.php:259 Stack trace: #0 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(906): Zend_Controller_Plugin_Broker->routeShutdown() #1 /usr/share/php/Icinga/Application/Web.php(471): Zend_Controller_Front->dispatch() #2 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}() #3 {main} thrown in /usr/share/icingaweb2/library/vendor/Zend/Controller/Plugin/Broker.php on line 259
```

Lots of deprecations also when running `icingacli`:
```
# icingacli help setup
PHP Deprecated:  Return type of Icinga\Application\Libraries::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Libraries.php on line 20

Deprecated: Return type of Icinga\Application\Libraries::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Libraries.php on line 20
PHP Deprecated:  Return type of Icinga\Application\Config::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 125

Deprecated: Return type of Icinga\Application\Config::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 125
PHP Deprecated:  Return type of Icinga\Application\Config::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 145

Deprecated: Return type of Icinga\Application\Config::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 145
PHP Deprecated:  Return type of Icinga\Application\Config::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 175

Deprecated: Return type of Icinga\Application\Config::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 175
PHP Deprecated:  Return type of Icinga\Application\Config::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 165

Deprecated: Return type of Icinga\Application\Config::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 165
PHP Deprecated:  Return type of Icinga\Application\Config::valid() should either be compatible with Iterator::valid(): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 155

Deprecated: Return type of Icinga\Application\Config::valid() should either be compatible with Iterator::valid(): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 155
PHP Deprecated:  Return type of Icinga\Application\Config::rewind() should either be compatible with Iterator::rewind(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 135

Deprecated: Return type of Icinga\Application\Config::rewind() should either be compatible with Iterator::rewind(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /usr/share/php/Icinga/Application/Config.php on line 135
PHP Deprecated:  preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /usr/share/php/Icinga/Application/Cli.php on line 126

Deprecated: preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated in /usr/share/php/Icinga/Application/Cli.php on line 126
PHP Deprecated:  preg_quote(): Passing null to parameter #1 ($str) of type string is deprecated in /usr/share/php/Icinga/Cli/Loader.php on line 279

Deprecated: preg_quote(): Passing null to parameter #1 ($str) of type string is deprecated in /usr/share/php/Icinga/Cli/Loader.php on line 279
USAGE: icingacli setup <command> [<action>] [options]

Available commands:

  config
  token

Show help on a specific command: icingacli help setup <command>
```

## To Reproduce
Provide a link to a live example, or an unambiguous set of steps to reproduce this issue. Include configuration, logs, etc. to reproduce, if relevant.

1. Install icingaweb2 Debian packages in an unstable environment
2. Add experimental apt sources
   ```
   echo "deb http://deb.debian.org/debian/ experimental main contrib non-free" >> /etc/apt/sources.list
   echo "deb-src http://deb.debian.org/debian/ experimental main contrib non-free" >> /etc/apt/sources.list
   ```
3. Install php8.1 packages from experimental: 
   ```
   apt update && apt install -t experimental libapache2-mod-php8.1 php8.1-{cli,curl,gd,intl,ldap,mbstring,mysql,opcache,pgsql,readline,soap,xml}
   ```
4. Enable mod_php8.1:
   ```
   a2enmod php8.1
   systemctl restart
   ```
5. `w3m http://localhost/icingaweb2`

## Expected behavior
Web interface works as expected

## Screenshots
N/A

## Your Environment
Include as many relevant details about the environment you experienced the problem in

* Icinga Web 2 version and modules (System - About): 2.9.5
* Web browser used: N/A
* Icinga 2 version used (`icinga2 --version`):
   ```
   icinga2 - The Icinga 2 network monitoring daemon (version: r2.13.2-1)
   
   Copyright (c) 2012-2021 Icinga GmbH (https://icinga.com/)
   License GPLv2+: GNU GPL version 2 or later <https://gnu.org/licenses/gpl2.html>
   This is free software: you are free to change and redistribute it.
   There is NO WARRANTY, to the extent permitted by law.
   
   System information:
     Platform: Debian GNU/Linux
     Platform version: Unknown
     Kernel: Linux
     Kernel version: 5.14.0-4-amd64
     Architecture: x86_64
   
   Build information:
     Compiler: GNU 11.2.0
     Build host: x86-ubc-02
     OpenSSL version: OpenSSL 1.1.1l  24 Aug 2021
   
   Application information:
   
   General paths:
     Config directory: /etc/icinga2
     Data directory: /var/lib/icinga2
     Log directory: /var/log/icinga2
     Cache directory: /var/cache/icinga2
     Spool directory: /var/spool/icinga2
     Run directory: /run/icinga2
   
   Old paths (deprecated):
     Installation root: /usr
     Sysconf directory: /etc
     Run directory (base): /run
     Local state directory: /var
   
   Internal paths:
     Package data directory: /usr/share/icinga2
     State path: /var/lib/icinga2/icinga2.state
     Modified attributes path: /var/lib/icinga2/modified-attributes.conf
     Objects path: /var/cache/icinga2/icinga2.debug
     Vars path: /var/cache/icinga2/icinga2.vars
     PID path: /run/icinga2/icinga2.pid
   ```
* PHP version used (`php --version`):
   ```
      PHP 8.1.0RC5 (cli) (built: Nov  4 2021 14:57:08) (NTS)
   Copyright (c) The PHP Group
   Zend Engine v4.1.0-dev, Copyright (c) Zend Technologies
       with Zend OPcache v8.1.0RC5, Copyright (c), by Zend Technologies
   ```
* Server operating system and version: Debian GNU/Linux bookworm/sid

## Additional context

 * https://github.com/php/php-src/blob/PHP-8.1/NEWS
 * https://github.com/php/php-src/blob/PHP-8.1/UPGRADING
 * https://github.com/php/php-src/blob/PHP-8.1/UPGRADING.INTERNALS

**Comment by nilmerg** [comment]
> Icinga\Application\Libraries::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in

So they finally gone nuts...

**Comment by Thomas-Gelf** [comment]
Depends on your point of view :rofl: Going with `#[\ReturnTypeWillChange]` is fine unless PHP 9. The checks for null (where not allowed for non-user functions) should have been addressed with 7.0, but this wasn't possible: at the time being nullable type declarations have still been missing. This is annoying, as similar problems can be hard to spot.

**Comment by nilmerg** [comment]
I wonder if we could postpone this until 2.11 with which we require at least PHP 7.2, then this *comment* won't be necessary to begin with :thinking:

**Comment by sebastic** [comment]
If 2.11 is planned for release reasonably soon (next three months or so) it could work, otherwise icingaweb2 will be unusable in testing & unstable once Debian transitions to php8.1. Fedora 36 uses 8.1 already.

**Comment by Thomas-Gelf** [comment]
> ...then this _comment_ won't be necessary to begin with

Not sure what you mean?! _Should have been addressed with 7.0 - if there where nullable types_ is what PHP people say about this issue, it is not related to Icinga Web. Missing nullable types in 7.0 were unfortunately on of the main reasons writing code requiring 7.0 was pointless.

**Comment by nilmerg** [comment]
> testing & unstable

Don't care about that.

> Fedora 36 uses 8.1 already

Hrg :neutral_face:

**Comment by Thomas-Gelf** [comment]
> ...otherwise icingaweb2 will be unusable in testing & unstable once Debian transitions to php8.1. Fedora 36 uses 8.1 already.

As a monitoring system, we run on many servers with "Enterprise" Linux Distributions. Icinga development is mostly sponsored by companies running them. Providing reliable support for old versions is therefore much more important than living on the bleeding edge. Even if the latter is much more fun.

Personally, I'm super excited about v8.1. Still, we have a large code-base that needs to support older versions for a long time to go.

Said all this: I do not really see a problem with supporting v8.1 pretty soon. Thanks to static analysis and powerful development environments, this could probably easily be tackled by an apprentice.

**Comment by sebastic** [comment]
If you don't keep up with changes in your dependencies, they will bite once their reach the next release of the Enterprise distributions you care about. RHEL 10 is going to have PHP > 8.1, Debian bookworm will too, as will Ubuntu 22.04.

**Comment by Thomas-Gelf** [comment]
Don't worry, we'll be ready ;-)

**Comment by Thomas-Gelf** [comment]
Related: fixes for [ipl\Html](https://github.com/Icinga/ipl-html/tree/fix/support-php8.1)

**Comment by Thomas-Gelf** [comment]
Short update on this issue: it's not a big deal. Sure, quite some work - and some of the problems are hard to spot. Most annoying: string functions which are now refusing `null`. Best example: `strlen($var)` used to be the best practice (and the fastest way) to check for "null or empty string" for a long time. Apart from this, our code base is quite clean and didn't cause many problems.

To solve this issue, we need new to release new versions for our libraries (ipl, third-party), the incubator and most of our modules. I moved my personal test systems from PHP 8 to PHP 8.1 and will now watch this for a couple of days. Once everything looks good, I'll finalize the pull requests.

### v2.9.5 with fixes for PHP 8.1

![icingaweb295_php81](https://user-images.githubusercontent.com/553008/143391047-5c7a1893-25a9-4c95-8c8b-ab4415f24cbd.png)

### The current Icinga Web 2 master branch

![icingaweb210pre_php81](https://user-images.githubusercontent.com/553008/143391068-530499c6-f9dc-4c49-a2d9-c9ca52a0e381.png)

**Comment by sebastic** [comment]
php8.1 is now in Debian testing & unstable, and icingaweb2 is scheduled for removal from Debian testing.

**Comment by csarn** [comment]
Ubuntu 22.04 LTS is released by now, which (as an LTS version) might also be a distribution that enterprises will want to run.
Installing icingaweb2 (2.9.5) from the repositories automatically installs php8.1 with it, and leads to the exact error this ticket is about.

**Comment by yhabteab** [comment]
```
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /icingaweb2/library/vendor/Zend/Mail.php on line 1224

PHP Deprecated:  str_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated in /icingaweb2/library/vendor/Zend/Mime.php on line 469
```

**Comment by HoffmannTom** [comment]
Will these fixes be available in Ubuntu 22 LTS Repository in future?

**Comment by nilmerg** [comment]
Of course they will, though there is no date yet when.

**Comment by Demonlinx** [comment]
So since there is no ETA to the fix of this bug it is marked as closed? Or just that it won't be fixed in 2.11.0?

**Comment by sebastic** [comment]
master has these changes for 8.1: https://github.com/Icinga/icingaweb2/pull/4799
